File: src/net/karlv/flang/parser/Lexer.scala
package net.karlv.flang.parser

import scala.util.parsing.combinator.RegexParsers
import java.io.InputStreamReader
import java.io.InputStream
import scala.util.parsing.input.CharSequenceReader

object Lexer extends RegexParsers {

  override def skipWhitespace = false;

  def tokens: Parser[List[Token]] =
    CharSequenceReader.EofCh ~> success(List(TEndOfInput)) |
    skip ~> tokens |
    token ~ tokens ^^ mkList;
  
  def apply(is: InputStream) = parseAll(tokens, new InputStreamReader(is)) match {
    case Success(ts, _) => ts;
    case failure: NoSuccess => sys.error(failure.toString());
  }
  
  def skip: Parser[Unit] = comment | ws;
  
  def token: Parser[Token] =
    (keywords ++ List(id, name, exprOp, int, float)).
      foldRight[Parser[Token]](failure("unexpected"))(_ | _);

  def keywords: List[Parser[TKeyword]] = List(
      "type", "val", "data", "sig", "(", ")", "open", "closed", "except", "only", "is", "rec", "let", "fn",
      "case", "begin", "do", "?", "where", "end", "in", "of", "module", "with", ":", "<:", ":>", "->", "<-",
      ";", "*", ".", "forall", "infix", "left", "right").map(_ ^^ TKeyword);

  def rawId: Parser[String] = """[a-zA-Z_][a-zA-Z0-9_]*""".r;
  
  def id: Parser[TId] = rawId ^^ TId;

  def name: Parser[TName] = rawId ~ rep("""\.""".r ~> rawId) ^^ mkList ^^ TName;

  def exprOp: Parser[TExprOp] = """(\+|-|\*|/|=|<|>)+""".r ^^ TExprOp;

  def int: Parser[TInt] = """[0-9]+""".r ^^ BigInt.apply ^^ TInt;

  def float: Parser[TFloat] =
    (("""[0-9]+\.[0-9]*""" + exponent).r | ("""\.[0-9]+""" + exponent).r | ("""[0-9]+""" + exponent).r) ^^
      BigDecimal.apply ^^ TFloat;

  def exponent = """[eE][+\-]?[0-9]+""";

  // TODO better comment syntax, with nestable block comments
  def comment: Parser[Unit] = """//[^\n\r]*\r?\n""".r ~> success(())

  def ws: Parser[Unit] = """[ \t\r\n]+""".r ~> success(());

  def charContent(quote: String, card: String) =
    ("""(\\[btnfr"'\\]|\\u[0-9a-fA-F]{4}|\\[0-3]?[0-7]{1,2}|[^\\""" + quote + "])" + card).r;

  def string: Parser[TString] = "\"" ~> charContent("\"", "*") <~ "\"" ^^ TString;

  def char: Parser[TChar] = "'" ~> charContent("'", "") <~ "'" ^^ TChar;

}

File: src/Lexer.hs

module Lexer (tokenize) where

import Text.Parsec hiding ((<|>), many, optional)
import Text.Parsec.String

import Common
import Token

tokenize :: String -> String -> FM [(Token, SourcePos)]
tokenize name xs = case parse file name xs of
  Left err -> fatal . Err ELexer Nothing Nothing . Just . show $ err
  Right ts -> return ts

withPos parser = do
  pos <- getPosition
  ret <- parser
  return (ret, pos)

file :: Parser [(Token, SourcePos)]
file = do
  skippable
  xs <- many1 (withPos tok) `sepEndBy` skippable
  eof
  return . concat $ xs

skippable = void . optional . many1 $ comment <|> (space >> return ())

tok = keywords <|> choice [ident, exprOp, litInt, litFloat, litString, litChar]

keywords =
  choice
  . map (try . (>>= return . TKeyword) . string)
  $ [ "type", "val", "data", "sig", "(", ")", "open", "closed", "except", "only", "is"
    , "rec", "let", "fn", "case", "if", "then", "else", "{", "}", "do", "in", "module"
    , "with", ":", "<:", ":>", "<-", ";", "forall", "infix" , "left", "right", "of"
    ]

ident = do
  h <- letter <|> char '_'
  t <- many $ alphaNum <|> char '_'
  return . TId . (:) h $ t

exprOp = many1 (oneOf "+-*/=:<>.?!@#$%^&|~") >>= return . TExprOp

litInt = many1 digit >>= return . TInt . read

litFloat :: Parser Token
litFloat = do
  (intPart, fracPart) <- withIntPart <|> withoutIntPart
  exp <- optionMaybe $ do
    _ <- oneOf "eE"
    sign <- optionMaybe $ oneOf "+-"
    let signVal = if sign == Just '-' then -1 else 1
    digs <- many1 digit
    return $ (signVal * read digs :: Integer)
  let intVal = read intPart :: Integer
  let fracVal = (read fracPart :: Integer) % (10 ^ length fracPart)
  return . TFloat $ (fromInteger intVal + fracVal) * (10 ^ maybe 0 id exp)
  where
    withIntPart = do
      intPart <- many1 digit
      _ <- char '.'
      fracPart <- many digit
      return (intPart, fracPart)
    withoutIntPart = do
      _ <- char '.'
      fracPart <- many1 digit
      return ("0", fracPart)

-- TODO better comment syntax, with nestable block comments
comment :: Parser ()
comment = do
  _ <- string "//"
  _ <- many . noneOf $ "\n\r"
  void . optional . char $ '\r'
  (char '\n' >> return ()) <|> eof

escapeCodes =
  map (\(e, r) -> (char e :: Parser Char) >> return r)
    [ ('b', '\b'), ('t', '\t'), ('n', '\n'), ('f', '\f'), ('r', '\r'), ('"', '"')
    , ('\'', '\''), ('\\', '\\')
    ]

charContent quote = (char '\\' >> escape) <|> normal
  where
    escape = foldr (<|>) (unicode <|> octal) escapeCodes
    unicode = do
      _ <- char 'u' :: Parser Char
      count 4 hexDigit >>= return . chr . read . ("0x" ++)
    octal = do
      a <- oneOf ['0' .. '3']
      [b, c] <- count 2 . oneOf $ ['0' .. '7']
      return
        . chr
        $ (readDigit a * 8 ^ (2 :: Integer)) + (readDigit b * 8) + readDigit c
    normal = noneOf [quote, '\\']
    readDigit = read . (: [])

litString = let q = char '"' in fmap TString . between q q . many . charContent $ '"'

litChar = let q = char '\'' in fmap TChar . between q q . charContent $ '\''

